ggml-vulkan.cpp

```diff
@@ -0,0 +1,5220 @@
+
+    s.wait_semaphores = std::move(wait_semaphores);
+    s.signal_semaphores = std::move(signal_semaphores);
+}
+
+static void ggml_vk_ctx_end(vk_context& ctx) {
+#ifdef VK_DEBUG
+    std::cerr << "ggml_vk_ctx_end(" << &ctx << ", " << ctx.seqs.size() << ")" << std::endl;
+#endif
+    if (ctx.s == nullptr) {
+        return;
+    }
+
+    ctx.s->buffer.end();
+    ctx.s = nullptr;
+}
+
+static void ggml_vk_ctx_begin(vk_context& ctx, vk_queue& q) {
+#ifdef VK_DEBUG
+    std::cerr << "ggml_vk_ctx_begin(" << &ctx << ")" << std::endl;
+#endif
+    if (ctx.s != nullptr) {
+        ggml_vk_ctx_end(ctx);
+    }
+
+    ctx.seqs.push_back({ ggml_vk_begin_submission(q) });
+    ctx.s = ctx.seqs[ctx.seqs.size() - 1].data();
+}
+
+static size_t ggml_vk_align_size(size_t width, size_t align) {
+    return CEIL_DIV(width, align) * align;
+}
+
+static void deferred_memcpy(void * dst, const void * src, size_t size, std::vector<vk_staging_memcpy>* memcpys = nullptr) {
+    if (memcpys == nullptr) {
+        memcpy(dst, src, size);
+    } else {
+        memcpys->emplace_back(dst, src, size);
+    }
+}
+
+static void ggml_vk_buffer_write_nc_async(vk_context& ctx, vk_buffer* dst, size_t offset, const ggml_tensor * tensor, vk_queue& q, std::vector<vk_staging_memcpy>* memcpys = nullptr, vk::Event * event = nullptr, bool allow_staging_resize = false) {
+#ifdef VK_DEBUG
+    std::cerr << "ggml_vk_buffer_write_nc_async(" << tensor << ")" << std::endl;
+#endif
+    GGML_ASSERT(!ggml_is_contiguous(tensor));
+    // Buffer is already mapped
+    if(dst->memory_property_flags & vk::MemoryPropertyFlagBits::eHostVisible) {
+        std::cerr << "ggml_vulkan: buffer_write_async dst buffer is host_visible. Use synchronous write." << std::endl;
+        GGML_ASSERT(false);
+    }
+    // Check if src is pinned memory
+    vk_buffer* buf = nullptr;
+    size_t buf_offset = 0;
+    for (size_t i = 0; i < vk_pinned_memory.size(); i++) {
+        const uint8_t* addr = (const uint8_t*) std::get<0>(vk_pinned_memory[i]);
+        const uint8_t* endr = addr + std::get<1>(vk_pinned_memory[i]);
+        if (tensor->data >= addr && tensor->data < endr) {
+            buf = &std::get<2>(vk_pinned_memory[i]);
+            buf_offset = ((const uint8_t *)tensor->data) - addr;
+            break;
+        }
+    }
+
+    const uint64_t ne0 = tensor->ne[0];
+    const uint64_t ne1 = tensor->ne[1];
+    const uint64_t ne2 = tensor->ne[2];
+    const uint64_t ne3 = tensor->ne[3];
+    const uint64_t nb0 = tensor->nb[0];
+    const uint64_t nb1 = tensor->nb[1];
+    const uint64_t nb2 = tensor->nb[2];
+    const uint64_t nb3 = tensor->nb[3];
+    const ggml_type type = tensor->type;
+    const uint64_t ts = ggml_type_size(type);
+    const uint64_t bs = ggml_blck_size(type);
+
+    const uint64_t dstnb0 = ts;
+    const uint64_t dstnb1 = dstnb0*(ne0/bs);
+    const uint64_t dstnb2 = dstnb1*ne1;
+    const uint64_t dstnb3 = dstnb2*ne2;
+
+    const uint64_t ne = ggml_nelements(tensor);
```

Comment:
These are all conflicting types which could cause implicit issues which are rooted in the warnings. This might be a minor concern considering the overall implementation.

```cpp
    const int64_t ne0 = tensor->ne[0];  // tensors are signed, so I updated from uint64_t
    const int64_t ne1 = tensor->ne[1];
    const int64_t ne2 = tensor->ne[2];
    const int64_t ne3 = tensor->ne[3];
    const int64_t nb0 = tensor->nb[0];
    const int64_t nb1 = tensor->nb[1];
    const int64_t nb2 = tensor->nb[2];
    const int64_t nb3 = tensor->nb[3];
    const ggml_type type = tensor->type;
    const size_t ts = ggml_type_size(type);  // this was defined as uint64_t, but returns size_t
    const int bs = ggml_blck_size(type);  // this was defined as uint64_t, but returns int


    const int64_t dstnb0 = ts;
    const int64_t dstnb1 = dstnb0 * static_cast<int64_t>(ne0 / bs);  // avoid implicitly casting to int64_t when assigning to dstnb
    const int64_t dstnb2 = dstnb1*ne1;
    const int64_t dstnb3 = dstnb2*ne2;

    const int64_t ne = ggml_nelements(tensor);  // was updated because this returns int64_t
```

Reply:
Nevermind. `vk::DeviceSize` requires `long unsigned int`.

Reply:
Yeah, I'm not sure why ggml uses signed integers for sizes and strides, since negative values don't make sense there. But it would be good to get rid of the warnings. I guess I'd have to explicitely cast everything? Or keep it signed and cast explicitely where it goes into the Vulkan API?

Reply:
@0cc4m

That's what I'm trying to figure out. It looks like we'll need a shim? Or at least an intermediary tensor, e.g. GGML:Vulkan compatible tensors. 

```cpp
    const ggml_type type = tensor->type;
    const size_t ts = ggml_type_size(type);
    const int bs = ggml_blck_size(type);

    const vk::DeviceSize vk_nelements = static_cast<vk::DeviceSize>(ggml_nelements(tensor));

    vk::DeviceSize ne[4] = {0, 0, 0, 0}; // Initialize with zeros or appropriate values
    vk::DeviceSize nb[4] = {0, 0, 0, 0}; // Initialize with zeros or appropriate values
    vk::DeviceSize dstnb[4] = {0, 0, 0, 0}; // Initialize with zeros or appropriate values

    for (int i = 0; i < 4; ++i) {
        ne[i] = static_cast<vk::DeviceSize>(tensor->ne[i]);
        nb[i] = static_cast<vk::DeviceSize>(tensor->nb[i]);
    }

    dstnb[0] = static_cast<vk::DeviceSize>(ts);
    dstnb[1] = dstnb[0] * static_cast<vk::DeviceSize>(ne[0] / bs);
    dstnb[2] = dstnb[1] * static_cast<vk::DeviceSize>(ne[1]);
    dstnb[3] = dstnb[2] * static_cast<vk::DeviceSize>(ne[2]);
```

Trying to figure out a sane way to implement this.

Reply:
Is there a reason not to simply cast everything to unsigned in the `ggml-vulkan.cpp` file? My compiler doesn't give warnings on lines like `const uint64_t ne00 = src0->ne[0];`, despite the different signedness. And I think we can assume that `ne` and `nb` will be positive.

Reply:
@0cc4m I get them all over the place.

```sh
g++ -I. -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG  -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -DGGML_USE_VULKAN -march=native -mtune=native -Wno-array-bounds -Wno-format-truncation -Wextra-semi -c ggml-vulkan.cpp -o ggml-vulkan.o
ggml-vulkan.cpp: In function ‘void ggml_vk_buffer_write_nc_async(vk_context&, vk_buffer*, size_t, const ggml_tensor*, vk_queue&, std::vector<vk_staging_memcpy>*, vk::Event*, bool)’:
ggml-vulkan.cpp:1400:33: warning: comparison of integer expressions of different signedness: ‘int64_t’ {aka ‘long int’} and ‘const uint64_t’ {aka ‘const long unsigned int’} [-Wsign-compare]
 1400 |         for (int64_t i3 = 0; i3 < ne3; i3++) {
      |   
```

This is just a snippet.

When comparing signed and unsigned integers, the signed integer is implicitly converted to an unsigned integer. This conversion can cause issues if the signed integer is negative, as it will convert to a large positive number due to the way negative numbers are represented in two's complement form. This can lead to logical errors in expressions/conditions.

We're implicitly converting values which affects the underlying representation. This ultimately has a cascading effect. While the range is technically the same, the way the information is represented differs. Typically the compiler handles all of this, but this is why type consistency matters.

You might think you have one value, but in reality you have another simply because the representation of that value changed unwittingly.

Reply:
I do get the warnings in loops and conditions, yes. I'll deal with those at the very least.

Reply:
The convention that is followed to a large extend in `ggml` is that iterators and tensor sizes in number of elements (`ne`) are `int64_t` while strides (`nb`) and memory offsets are `size_t`. The reasoning is that indexes are often used in subtractions and thus signed int avoids overflows

It's not mandatory, but it would be nice to follow this convention whenever possible mainly for consistency. The `GGML_TENSOR_LOCAL` macros can be useful to handle dimensions and strides